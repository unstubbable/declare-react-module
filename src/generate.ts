import {generateFromSource} from 'react-to-typescript-definitions';
import {rollup} from 'rollup';
import jsx = require('rollup-plugin-jsx');
import resolve = require('rollup-plugin-node-resolve');
import {
  replaceWithEmptyModule
} from './rollup-plugin-replace-with-empty-module';

async function generateBundleCode(entry: string): Promise<string> {
  const bundle = await rollup({
    entry,
    plugins: [
      replaceWithEmptyModule(['**/*.css']),
      resolve({
        extensions: ['.js', '.jsx']
      }),
      jsx({factory: 'React.createElement'})
    ]
  });
  return bundle.generate({format: 'es'}).code;
}

export async function generateTypings(
  moduleName: string,
  entry: string
): Promise<string> {
  const code = await generateBundleCode(entry);
  return generateFromSource(moduleName, code);
}
