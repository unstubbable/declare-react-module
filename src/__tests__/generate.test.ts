import {generateTypings} from '../generate';
import {getInputFileName} from './helper';

function itGeneratesTypingsForModule(name: string): void {
  it(`generates typings for module ${name}`, async () => {
    const input = getInputFileName(name);
    const typings = await generateTypings(name, input);

    expect(typings).not.toMatch('\r\n');
    expect(typings).toMatchSnapshot();
  });
}

itGeneratesTypingsForModule('export-from-js');
itGeneratesTypingsForModule('export-from-jsx');
itGeneratesTypingsForModule('stage-2-syntax');
itGeneratesTypingsForModule('with-css-modules');
itGeneratesTypingsForModule('with-dependency');
itGeneratesTypingsForModule('with-es5-dependency');
itGeneratesTypingsForModule('with-json-dependency');
itGeneratesTypingsForModule('with-node-builtin-dependency');
