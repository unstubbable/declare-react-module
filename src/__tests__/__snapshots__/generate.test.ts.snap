// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates typings for module export-from-js 1`] = `
"declare module 'export-from-js' {
    export interface ComponentProps {
        foo: string;
    }

    export function Component(props: ComponentProps): JSX.Element;

}

"
`;

exports[`generates typings for module export-from-jsx 1`] = `
"declare module 'export-from-jsx' {
    export interface ComponentProps {
        foo: string;
    }

    export function Component(props: ComponentProps): JSX.Element;

}

"
`;
